src/python/pants/backend/shell/util_rules/shell_command.py

```diff
@@ -276,6 +294,90 @@ async def run_shell_command(
     return GeneratedSources(output)
 
 
+@rule(desc="Running run_in_sandbox target", level=LogLevel.DEBUG)
+async def run_in_sandbox_request(
+    request: GenerateFilesFromRunInSandboxRequest,
+) -> GeneratedSources:
+    shell_command = request.protocol_target
+    description = f"the `{shell_command.alias}` at {shell_command.address}"
+    environment_name = await Get(
+        EnvironmentName, EnvironmentNameRequest, EnvironmentNameRequest.from_target(shell_command)
+    )
+
+    runnable_address_str = shell_command[RunInSandboxRunnableField].value
+    if not runnable_address_str:
+        raise Exception(f"Must supply a `runnable` for {description}.")
+
+    runnable_address = await Get(
+        Address,
+        AddressInput,
+        AddressInput.parse(
+            runnable_address_str,
+            relative_to=shell_command.address.spec_path,
+            description_of_origin=f"The `{RunInSandboxRunnableField.alias}` field of {description}",
+        ),
+    )
+
+    addresses = Addresses((runnable_address,))
+    addresses.expect_single()
+
+    runnable_targets = await Get(Targets, Addresses, addresses)
+    field_sets = await Get(
+        FieldSetsPerTarget, FieldSetsPerTargetRequest(RunFieldSet, runnable_targets)
+    )
+    run_field_set: RunFieldSet = field_sets.field_sets[0]
+
+    working_directory = shell_command.address.spec_path
+
+    # Must be run in target environment so that the binaries/envvars match the execution
+    # environment when we actually run the process.
+    run_request = await Get(
+        RunInSandboxRequest, {environment_name: EnvironmentName, run_field_set: RunFieldSet}
+    )
+
+    dependencies_digest = await _execution_environment_from_dependencies(shell_command)
+
+    input_digest = await Get(Digest, MergeDigests((dependencies_digest, run_request.digest)))
+
+    output_files, output_directories = _parse_outputs_from_command(shell_command, description)
+
+    extra_args = shell_command.get(RunInSandboxArgumentsField).value or ()
```

Comment:
Doesn't `.get()` allow you to provide a default value?

```suggestion
    extra_args = shell_command.get(RunInSandboxArgumentsField, ()).value
```

Reply:
@thejcannon AIUI, that's for when the `Field` is not provided at all. It allows you to provide the positional argument `default_raw_value=`, which will create an entire new `Field` instance with the default value provided. It's a lot more typey-type than this when you definitely have a field instance available.

In this case, the field is defined with a default of `()` and the `or ()` is there to ingratiate `mypy`

Reply:
Oh yeah duh!